torch/csrc/lazy/core/dynamic_ir.h

```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+#include <ATen/core/symbol.h>
+
+#include <functional>
+#include <memory>
+#include <set>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include <c10/core/ScalarType.h>
+#include <torch/csrc/lazy/core/hash.h>
+#include <torch/csrc/lazy/core/ir.h>
+#include <torch/csrc/lazy/core/ir_metadata.h>
+#include <torch/csrc/lazy/ts_backend/ts_node.h>
+#include <c10/util/Flags.h>
+
+C10_DECLARE_bool(ltc_enable_dynamic_shapes);
+
+namespace torch {
+namespace lazy {
+
+/**
+ * The goal of "dynamic" Nodes is to patch a hole in our tracing.
+ * Previously, if a user called `sizes` on a Tensor, it would leak out
+ * of our tracing system, as `sizes` returns a torch.Size or an int. To
+ * prevent this from happening, we introduce DimensionNode, a new type
+ * of Node that abstracts the operation of getting the dimensions of a
+ * Tensor.
+ * 
+ * Consider the following example:
+ * ```
+ * numel = x.shape()[0] * x.shape()[1]
+ * ```
+ * 
+ * Here, `x.shape()[i]` will be a SizeNode (subclass of DimensionNode),
+ * and the multiplication of the two SizeNodes will be represented by
+ * a SizeMul (also a subclass of DimensionNode). Through this, we can
+ * prevent `numel` from being represented as a Python int and thus
+ * burned into the Graph.
+ */
+
+class TORCH_API DimensionNode : public lazy::TsNode {
```

Comment:
A couple of comments RE the base class.

1) files in lazy/core should probably derive from Node, not TsNode
2) DimensionNode and related nodes will be constructed by LazyTensor sizes() method, and Xla already has a plan for overriding LazyTensor with XlaTensor, which gives a hook point for constructing Xla versions of these nodes

For now, I think the baseline assumption is that these classes are not literally usable by XLA, but they would have to copy them;  hence moving them under ts_backend folder is more correct.

But i'm in favor of landing this as-is, since there are other nodes in lazy/core which are derived from TsNode already, and we have a separate effort to disentangle them.  It's probably best to just land this and then clean up all the nodes at once.

cc @miladm @JackCaoG

Reply:
what's the blocker for making `DimensionNode` inherited from `node` instead of `tsNode`? Is it the lower function?

Reply:
no, there is no blocker really.  

other than, we need to discuss the general design of nodes and inheritance :)

I have been operating on the assumption that
 - xla needs to have 'XlaNode' be the base class for all (only some?) of the nodes
 - it's much simpler if the backend can assume every node is derived from BackendNode, than if the backend has to check each node in the graph and it could either be e.g. XlaNode or Node
 
 We should hash this part out along with @antoniojkim's proposal for IrBuilder and conclude on the direction
 
From Nick's perspective, we could just make DimensionNode :public Node today.  We can drop the 'Lower' method from being an IR method and instead codegen a standalone 'chain-of-if-statement' version, and totally get rid of TsNode if we want to.  Part of the reason I kept TsNode around is to make sure it's possible for you to keep XlaNode around.

Reply:
oh right. Yea I think it make sense for `DimensionNode` to inherited from `tsNode` now. PyTorch/XLA code will have an assumption that all of the input node is a `XlaNode` and have `xla_shape`. In the short term let's keep it like this.

Reply:
`Shape`s don't really make much sense for `DimensionNode` tbh. I'm actually thinking if `DimensionNode` should explicitly override `shape()` to throw an error if some code erroneously try to access it.
What do you guys think?

Reply:
> `Shape`s don't really make much sense for `DimensionNode` tbh. I'm actually thinking if `DimensionNode` should explicitly override `shape()` to throw an error if some code erroneously try to access it. What do you guys think?

yea i pretty much agree.  i think the fork point is "our IR implicitly is always typed as Tensor"
 - true: we either make 'shape()' raise an error (which exposes that we violate this implicit type), or we return shape {}, which kind of blurs a bit that a dimension is like a scalar-tensor
 - false: ok, so we finally have to add types to our IR... but that's not a decision to make rapidly

Reply:
+1 for shape to throw an error.

Reply:
This is a great question @wconstab. I need to give this some more thought and circle back. 

My immediate thoughts: 
What we ultimately need is to convert the IR node from the `SymInt` input to XLAOp and pass it down for lowering. At lowering, I can think of scenarios where we want to query the shape of the XLAOp. For instance, here is how I am thinking we utilize `target_size` while lowering `expand`:

```
xla::XlaOp BuildDynamicExpand(xla::XlaOp input,
                              xla::XlaOp target_size) {

  xla::Shape output_sizes = XlaHelpers::ShapeOfXlaOp(target_size);
  xla::XlaOp output = BuildExpand(input, output_sizes.dimensions());

  for (int i = 0; i < output_sizes.size(); ++i) {
    if (output_sizes.is_dynamic_dimension(i)) {
      output = xla::SetDimensionSize(output, target_size, i);
    }
  }
  return output;
}
```

What do you think?